server/otlp/http_server.go

```diff
@@ -0,0 +1,222 @@
+	"google.golang.org/protobuf/proto"
+)
+
+const protoBufContentType = "application/x-protobuf"
+const jsonContentType = "application/json"
+const gzipEncoding = "gzip"
+
+type HttpServer struct {
+	addr     string
+	exporter IExporter
+
+	hServer *http.Server
+}
+
+func NewHttpServer(addr string, exporter IExporter) *HttpServer {
+	return &HttpServer{
+		addr:     addr,
+		exporter: exporter,
+	}
+}
+
+func (s *HttpServer) Start() error {
+	s.hServer = &http.Server{
+		Addr:    s.addr,
+		Handler: s,
+	}
+
+	listener, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return fmt.Errorf("cannot listen on address %s: %w", s.addr, err)
+	}
+
+	return s.hServer.Serve(listener)
+}
+
+func (s *HttpServer) Stop() {}
+
+func (s HttpServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	contentType, contentEncoding, err := s.validateRequest(r)
+	shouldDecompress := contentEncoding == gzipEncoding
+	shouldEncode := strings.Contains(r.Header.Get("Accept-Encoding"), gzipEncoding)
+
+	response := newHttpResponse(w, contentType, shouldEncode)
+	if err != nil {
+		response.sendError(http.StatusBadRequest, status.Errorf(codes.InvalidArgument, "Error when validating request %s", err.Error()))
+		return
+	}
+
+	request, err := s.parseBody(r.Body, contentType, shouldDecompress)
+	if err != nil {
+		response.sendError(http.StatusUnprocessableEntity, status.Errorf(codes.InvalidArgument, "Could not parse request body %s", err.Error()))
+		return
+	}
+
+	result, err := s.exporter.Ingest(r.Context(), request)
+	if err != nil {
+		response.sendError(http.StatusInternalServerError, status.Errorf(codes.InvalidArgument, "Error when ingesting spans %s", err.Error()))
+		return
+	}
+	response.send(http.StatusOK, result)
+}
+
+func (s HttpServer) parseProtoBuf(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	request := pb.ExportTraceServiceRequest{}
+
+	err := proto.Unmarshal(body, &request)
+	if err != nil {
+		return nil, err
+	}
+
+	return &request, nil
+}
+
+func (s HttpServer) parseJson(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	request := pb.ExportTraceServiceRequest{}
+	err := json.Unmarshal(body, &request)
+	if err != nil {
+		return nil, err
+	}
+
+	return &request, nil
+}
+
+func (s HttpServer) parseBody(reqBody io.ReadCloser, contentType string, shouldDecompress bool) (*pb.ExportTraceServiceRequest, error) {
+	var body []byte
+	if b, err := io.ReadAll(reqBody); err == nil {
+		body = b
+	} else {
+		return nil, err
+	}
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("empty body")
+	}
+
+	if shouldDecompress {
+		decompressedBody, err := s.decompressBody(body)
+		if err != nil {
+			return nil, err
+		}
+
+		body = decompressedBody
+	}
+
+	switch contentType {
+	case protoBufContentType:
+		return s.parseProtoBuf(body)
+	case jsonContentType:
+		return s.parseJson(body)
+	}
+
+	return nil, fmt.Errorf("content-type %s not supported", contentType)
+}
+
+func (s HttpServer) validateRequest(r *http.Request) (string, string, error) {
```

Comment:
most of this could be replaced by using gorilla mux and define a route. It might look overkill but it's A LOT less code to maintain, supports the gzip middleware out of the box, and we are already using it, so we keep our "http services" consistent across packages

Reply:
done